add metrics for udp received/dropped/error

mieczkowski opened the pull request:
I've added metrics for receiving packets:
1. Received
2. Dropped (important for our anty-ddos project)
3. Invalid & error

This is easy version - just registering metrics in udp.go.
Alternative: separate metrics from metrics/metrics.go to package, because in metrics/decoder.go there is import for "utils" pkg, so in utils we can't use metrics. And I don't know if this is preferred.

![image](https://github.com/netsampler/goflow2/assets/3045505/22ea56f4-97a6-481f-a55f-eccc448e8002)

mieczkowski commented:
So maybe second option: move PromProducerWrapper PromTemplateSystem and PromDecoderWrapper to separate package (packages?), and leave only counters in metrics package (to be used in every package in project)

lspgn commented:
One more thing:
[`MetricTrafficPackets` ](https://github.com/netsampler/goflow2/blob/8428ddad91238ff8f520a0291c87a069fd7ad3ce/metrics/metrics.go#L20) used in [`decoder.go`](https://github.com/netsampler/goflow2/blob/8428ddad91238ff8f520a0291c87a069fd7ad3ce/metrics/decoder.go#L34) gives the same information as `MetricUdpPacketReceived`.

mieczkowski commented:
> MetricTrafficPackets used in decoder.go gives the same information as MetricUdpPacketReceived.

Right, removed, and added _total suffix (promlinter)

> It could be interesting to have subpackages for decoder, producer and template but that shouldn't serve so metrics can be imported to avoid cycles. I designed it so metrics are defined at the top level and a user can define their own metric system.

If metrics are global for project they should be in separate package, so every package could use them. In this solution - package utils is "discriminated" against using metrics, because... metrics need utils package 🤣  So utils package must declare metrics in way like I did. 

And this dropped_packets metrics is important, because I didn't even noticed that in high volume of traffic I was loosing data (only by debugging code)

> have you thought about using blocking mode

With nonblocking mode and metrics I can be sure that goflow2 is dropping packets (and for example: increment workers or sockets value). With kernel dropped packets I'm not sure how many packets are dropped in goflow2 (if there is more services on host)

lspgn commented:
I'll think over the weekend. Again, metrics should not be defined outside the metrics package as I don't want to force importers to rely on Prometheus and increase a counter that may not be used. This can be solved with a callback mechanism. Furthermore, labels are inconsistent.
Keep in mind GoFlow2 is imported in various tools.

> With nonblocking mode and metrics I can be sure that goflow2 is dropping packets (and for example: increment workers or sockets value). With kernel dropped packets I'm not sure how many packets are dropped in goflow2 (if there is more services on host)

Please note that you can also have kernel dropped packets in a non-blocking scenario too.
Additionally, I disagree on surfacing the source address and port for dropped and invalid. This could easily cause a cardinality explosion.

lspgn commented:
Started tackling it with callbacks on #226

mieczkowski commented:
Ok I will look later. 

Btw with metrics I've added: 

![image](https://github.com/netsampler/goflow2/assets/3045505/599c752c-5d0b-4d72-b65c-66d795975e2c)

There was already one moment when packets was dropped, even with settings:

```
sflow://:6343?count=2&workers=4&queue_size=100000
```

So now I know that I should increase count and/or workers ;) 

> Please note that you can also have kernel dropped packets in a non-blocking scenario too.

No with bare metal instance with another services on it :(

lspgn commented:
Hi @mieczkowski 
Have you had a change to look at the PR?

Thank you :)

mieczkowski commented:
Oh, I forgot, too much work :( I will run and test it tomorrow max :)
